### PR TITLE
Azure SD: check for nil before accessing and dereferencing 

### PR DESCRIPTION
**Pull request by darshanime**
- check for nil before accessing and dereferencing 
- add test for above, remove code duplication

closes https://github.com/prometheus/prometheus/issues/13574

### CODE REVIEW

**Comment by Nexucis** [comment]
ah that's a serious bugfix here.
Let's add it to the 2.50. Can you @darshanime change the target branch to `release-2.50` please ?

**Comment by darshanime** [comment]
done @Nexucis, ptal

File: discovery/azure/azure.go [hunk 1]
```diff
@@ -569,7 +569,13 @@ func (client *azureClient) getScaleSetVMs(ctx context.Context, scaleSet armcompu
 }
 
 func mapFromVM(vm armcompute.VirtualMachine) virtualMachine {
-	osType := string(*vm.Properties.StorageProfile.OSDisk.OSType)
+	var osType string
+	if vm.Properties != nil &&
```
Comment (bboreham) [inline discovery/azure/azure.go:573]: Move this inside the existing `if vm.Properties != nil` (previously on line 582)

File: discovery/azure/azure.go [hunk 2]
```diff
@@ -608,7 +614,13 @@ func mapFromVM(vm armcompute.VirtualMachine) virtualMachine {
 }
 
 func mapFromVMScaleSetVM(vm armcompute.VirtualMachineScaleSetVM, scaleSetName string) virtualMachine {
-	osType := string(*vm.Properties.StorageProfile.OSDisk.OSType)
+	var osType string
+	if vm.Properties != nil &&
```
Comment (bboreham) [inline discovery/azure/azure.go:618]: Move this inside the existing `if vm.Properties != nil` (previously on line 621)

**Review by bboreham — COMMENTED** [review]
Thanks for this; one simplification suggested.

It would be easier to review the two commits as two PRs.

**Review by bboreham — APPROVED** [review]
lgtm

**Pull request merged**